PxMatrix.cpp

```diff
@@ -0,0 +1,955 @@
+    // the same value to r1/r2/g1/g2/b1/b2 at the same time to load the config into all the FM6126 chips
+
+    // Some necessary magic bit fields
+    // b12  - 1  adds red tinge
+    // b12  - 9/8/7/6/5  =  4 bit brightness
+    // b13  - 9   =1 screen on
+    // b13  - 6   =1 screen off
+    pinMode(_SPI_CLK, OUTPUT);
+    pinMode(_SPI_MOSI, OUTPUT);
+    digitalWrite(_SPI_CLK, HIGH);  // CCK LOW
+    digitalWrite(_OE_PIN, LOW);
+    digitalWrite(_LATCH_PIN, HIGH);
+    digitalWrite(_A_PIN, HIGH);
+    digitalWrite(_B_PIN, LOW);
+    digitalWrite(_C_PIN, LOW);
+    digitalWrite(_D_PIN, LOW);
+
+    uint8_t reg_bit = 0;
+    for (uint32_t bit_counter = 0; bit_counter < _send_buffer_size * 8; bit_counter++) {
+      reg_bit = bit_counter % 16;
+      if ((reg_value >> reg_bit) & 1)
+        digitalWrite(_SPI_MOSI, HIGH);
+      else
+        digitalWrite(_SPI_MOSI, LOW);
+
+      delay(1);
+      digitalWrite(_SPI_CLK, LOW);  // CLK HIGH
+      delay(1);
+      digitalWrite(_SPI_CLK, HIGH);  // CLK LOW
+      delay(1);
+      if ((bit_counter == (_send_buffer_size * 8 - reg_position - 1))) {
+        digitalWrite(_LATCH_PIN, LOW);
+      }
+    }
+    digitalWrite(_LATCH_PIN, HIGH);
+  }
+  digitalWrite(_OE_PIN, HIGH);
+}
+#endif
+
+void PxMATRIX::setDriverChip(driver_chips driver_chip) {
+  _driver_chip = driver_chip;
+
+  if (driver_chip == FM6124 || driver_chip == FM6126A) {
+    uint16_t b12a = 0b0111111111111111;  // 亮度: high
+    b12a = 0b0111100011111111;           // 亮度: low
+    // uint16_t b12b=0b0111100000111111;
+    // uint16_t b12c=0b0111111111111111;
+    // uint16_t b12d=0b0111100000111111;
+
+    uint16_t b13a = 0b0000000001000000;
+    // uint16_t b13b=0b0000000001000000;
+    // uint16_t b13c=0b0000000001000000;
+    // uint16_t b13d=0b0000000001000000;
+
+#ifdef ESP32
+    pinMode(_OE_PIN, OUTPUT);
+    pinMode(_LATCH_PIN, OUTPUT);
+    digitalWrite(_OE_PIN, HIGH);
+    pinMode(_LATCH_PIN, LOW);
+
+    fm612xWriteRegister(b12a, 11);
+    fm612xWriteRegister(b13a, 12);
+
+#else
+    writeRegister(b12a, 12);
+    writeRegister(b13a, 13);
+#endif
+  }
+}
+
+void PxMATRIX::setMuxPattern(mux_patterns mux_pattern) {
+  _mux_pattern = mux_pattern;
+
+  // We handle the multiplexing in the library and activate one of for
+  // row drivers --> need A,B,C,D pins
+  if (_mux_pattern == STRAIGHT) {
+    pinMode(_C_PIN, OUTPUT);
+    pinMode(_D_PIN, OUTPUT);
+  }
+}
+
+void PxMATRIX::setColorOrder(color_orders color_order) { _color_order = color_order; }
+
+void PxMATRIX::setMuxDelay(uint8_t mux_delay_A, uint8_t mux_delay_B, uint8_t mux_delay_C, uint8_t mux_delay_D,
+                           uint8_t mux_delay_E) {
+  _mux_delay_A = mux_delay_A;
+  _mux_delay_B = mux_delay_B;
+  _mux_delay_C = mux_delay_C;
+  _mux_delay_D = mux_delay_D;
+  _mux_delay_E = mux_delay_E;
+}
+
+void PxMATRIX::setScanPattern(scan_patterns scan_pattern) { _scan_pattern = scan_pattern; }
```

Comment:
oneliners like this probably violate our code style. i don't really mind though.